config/initializers/oracle.rb

```diff
@@ -52,13 +52,32 @@ def _quote(value)
             super
           end
         end
+
+        # Fixing OCIError: ORA-01741: illegal zero-length identifier
+        # because of https://github.com/rails/rails/commit/c18a95e38e9860953236aed94c1bfb877fa3be84
+        # the value of `columns` is  [ "\"ACCOUNTS\".\"ID\"" ] which forms an incorrect query
+        # ... OVER (PARTITION BY ["\"ACCOUNTS\".\"ID\""] ORDER BY "ACCOUNTS"."ID") ...
+        def columns_for_distinct(columns, orders) # :nodoc:
+          # construct a valid columns name for DISTINCT clause,
+          # ie. one that includes the ORDER BY columns, using FIRST_VALUE such that
+          # the inclusion of these columns doesn't invalidate the DISTINCT
+          #
+          # It does not construct DISTINCT clause. Just return column names for distinct.
+          orders.reject(&:blank?).map { |s|
+            s = visitor.compile(s) unless s.is_a?(String)
+            # remove any ASC/DESC modifiers
+            s.gsub(/\s+(ASC|DESC)\s*?/i, "")
+          }.reject(&:blank?).map.with_index { |column, i|
+            "FIRST_VALUE(#{column}) OVER (PARTITION BY #{columns.join(', ')} ORDER BY #{column}) AS alias_#{i}__"
```

Comment:
The original implementation is:

```
        order_columns = orders.reject(&:blank?).map { |s|
            s = visitor.compile(s) unless s.is_a?(String)
            # remove any ASC/DESC modifiers
            s.gsub(/\s+(ASC|DESC)\s*?/i, "")
          }.reject(&:blank?).map.with_index { |column, i|
            "FIRST_VALUE(#{column}) OVER (PARTITION BY #{columns} ORDER BY #{column}) AS alias_#{i}__"
          }
        (order_columns << super).join(", ")
```

But using `super` was causing a bad query too.

Reply:
Are we sure `super` would not add needed columns sometimes?

Just wondering. This code path doesn't seem very common.

One way to make `super` work would be to add to the  anonymous module definition this:
```ruby
    extend ActiveSupport::Concern

    included do
      remove_method :columns_for_distinct
    end
``` 

In this way, `super` will reach whatever is underneath `ActiveRecord::ConnectionAdapters::OracleEnhancedAdapter` instead of getting into an infinite recursion.

Reply:
it seems to basically add possible other columns separated by comma to the other query.

Digging into the code I didn't see how other columns may appear there though.

Also I wont be surprised in case adding columns with comma will break the SQL.

I assume we can test that by adding some artificial column and see. The issue is that I couldn't identify the precise condition that this code is reached. And it is hard to justify spending more time on it.

So I guess we are good to go as is I assume.